wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -305,26 +324,53 @@ public void stop(String workspaceId, Map<String, String> options)
       throw new ConflictException(
           format("Could not stop workspace '%s' because its state is '%s'", workspaceId, status));
     }
+    String stoppedBy =
+        firstNonNull(
+            sessionUserNameOr(workspace.getAttributes().get(WORKSPACE_STOPPED_BY)), "undefined");
+    LOG.info(
+        "Workspace '{}/{}' with id '{}' is being stopped by user '{}'",
+        workspace.getNamespace(),
+        workspace.getConfig().getName(),
+        workspace.getId(),
+        stoppedBy);
     eventService.publish(
         DtoFactory.newDto(WorkspaceStatusEvent.class)
             .withWorkspaceId(workspaceId)
             .withPrevStatus(state.status)
             .withStatus(STOPPING));
 
-    try {
-      state.runtime.stop(options);
-
-      // remove before firing an event to have consistency between state and the event
-      runtimes.remove(workspaceId);
-      publishWorkspaceStoppedEvent(workspaceId, STOPPING, null);
-    } catch (InfrastructureException e) {
-      // remove before firing an event to have consistency between state and the event
-      runtimes.remove(workspaceId);
-      publishWorkspaceStoppedEvent(
-          workspaceId,
-          STOPPING,
-          "Error occurs on workspace runtime stop. Error: " + e.getMessage());
-    }
+    return CompletableFuture.runAsync(
+        ThreadLocalPropagateContext.wrap(
+            () -> {
+              try {
+                state.runtime.stop(options);
+
+                // remove before firing an event to have consistency between state and the event
+                runtimes.remove(workspaceId);
+                LOG.info(
+                    "Workspace '{}/{}' with id '{}' stopped by user '{}'",
+                    workspace.getNamespace(),
+                    workspace.getConfig().getName(),
+                    workspace.getId(),
+                    stoppedBy);
+                publishWorkspaceStoppedEvent(workspaceId, STOPPING, null);
+              } catch (InfrastructureException e) {
+                // remove before firing an event to have consistency between state and the event
+                runtimes.remove(workspaceId);
+                LOG.info(
+                    "Workspace '{}/{}' with id '{}' stopped by user '{}'. Error: {}",
+                    workspace.getNamespace(),
+                    workspace.getConfig().getName(),
+                    workspace.getId(),
+                    stoppedBy,
+                    e);
```

Comment:
Are you sure that `Logger` will process exception object correctly? Mb it'd be better to log `e.getMessage`?